addok_france/utils.py

```diff
@@ -37,11 +39,45 @@ def extract_address(q):
     flags=re.IGNORECASE)
 
 
-def glue_ordinal(q):
+def neighborhood(iterable, first=None, last=None):
+    """
+    Yield the (previous, current, next) items given an iterable.
+
+    You can specify a `first` and/or `last` item for bounds.
+    """
+    iterator = iter(iterable)
+    previous = first
+    current = next(iterator)  # Throws StopIteration if empty.
+    for next_ in iterator:
+        yield (previous, current, next_)
+        previous = current
+        current = next_
+    yield (previous, current, last)
+
+
+def glue_ordinal(tokens):
     """Glue '3' and 'bis'."""
-    return glue_ordinal_pattern.sub('\g<1>\g<2>\g<3>', q)
-glue_ordinal_pattern = re.compile('(\d{1,4}) (' + ORDINAL_REGEX + ')\\b($|(?:,? (' + TYPES_REGEX + ')))',  # noqa
-                                  flags=re.IGNORECASE)
+    previous = None
+    for _, token, next_ in neighborhood(tokens):
+        if token.isdigit():
+            previous = token
+        elif previous is not None:
+            # Matches "bis" either followed by a type or nothing.
+            if (ordinal_pattern.match(token) and
+                    (not next_ or (next_ and types_pattern.match(next_)))):
+                raw = '{} {}'.format(previous, token)
+                # Space removed to maximize chances to get a hit.
+                token = token.update(raw.replace(' ', ''), raw=raw)
+            else:
+                yield previous
+            previous = None
+            yield token
```

Comment:
WIth a `continue` in the first `if` we can remove one of the two `yield token` I think